src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py

```diff
@@ -12,3 +40,198 @@ class InTotoV1Statement(TypedDict):
+        )
+
+    name_valid = _validate_property(subject, "name", False, lambda x: isinstance(x, str))
+    if not name_valid:
+        raise ValidateInTotoPayloadError("The subject name is not valid.")
+
+    # At least one of 'uri', 'digest', and 'content' must be valid and present.
+    uri_valid = _validate_property(subject, "uri", True, lambda x: isinstance(x, str))
+    content_valid = _validate_property(subject, "content", True, lambda x: isinstance(x, str))
+    digest_valid = _validate_property(subject, "digest", True, is_valid_digest_set)
+    if not (uri_valid or content_valid or digest_valid):
+        raise ValidateInTotoPayloadError(
+            "One of 'uri', 'digest', or 'content' must be present and valid within 'subject'."
+        )
+
+    download_location_valid = _validate_property(subject, "downloadLocation", False, lambda x: isinstance(x, str))
+    if not download_location_valid:
+        raise ValidateInTotoPayloadError("The subject downloadLocation is not valid.")
+
+    media_type_valid = _validate_property(subject, "mediaType", False, lambda x: isinstance(x, str))
+    if not media_type_valid:
+        raise ValidateInTotoPayloadError("The subject mediaType is not valid.")
+
+    annotations_valid = _validate_property(subject, "annotations", False, is_valid_annotation_map)
+    if not annotations_valid:
+        raise ValidateInTotoPayloadError("The subject annotations are not valid.")
+
+    return True
+
+
+def is_valid_digest_set(digest: JsonType) -> bool:
+    """Validate the digest set.
+
+    Specification for the digest set: https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md.
+
+    Parameters
+    ----------
+    digest : JsonType
+        The digest set.
+
+    Returns
+    -------
+    bool:
+        ``True`` if the digest is valid according to the spec.
+    """
+    if not isinstance(digest, dict):
+        return False
+    for key in digest:
+        if key not in VALID_ALGORITHMS:
+            return False
+        if not isinstance(digest[key], str):
+            return False
+    return True
+
+
+def is_valid_annotation_map(annotation_map: JsonType) -> bool:
+    """Validate the annotation map which is a dictionary with string keys and JsonType values.
+
+    Parameters
+    ----------
+    annotation_map : JsonType
+        The annotation map dictionary.
+
+    Returns
+    -------
+    bool:
+        ``True`` if the annotation map is valid according to the spec.
+    """
+    if not isinstance(annotation_map, dict):
+        return False
+    return True
+
+
+def _validate_property(
+    object_: JsonType, target_name: str, required: bool, validator_function: Callable[[JsonType], bool]
+) -> bool:
+    """Validate the existence and type of target within the passed Json object."""
+    if not isinstance(object_, dict):
+        return False
+
+    target = object_.get(target_name)
+    if not target:
+        return not required
+
+    return validator_function(target)
```

Comment:
Nit: `property_name` (or `key`) and `value` would be a more familiar naming scheme for me than `target_name` and `target`. Also in that case, `validator_function` can be called `value_validator`, since it is clearer what is being validated, and the word `function` may not be as necessary since it is already type-annotated as a `Callable`. What do you think?